### DISCUSSION TITLE
`encrypted` option doco unclear - no way to send filesystems unchanged?

### DISCUSSION THREAD

**Discussion by Malvineous** [zrepl/zrepl#857] [Q&A] [answered]
Hi all,

Does anyone know how the `encrypted` option works?  It's listed in the [documentation](https://zrepl.github.io/configuration/sendrecvoptions.html#encrypted) but all it says is `encrypted=true` only sends encrypted file systems, and `encrypted=false` expects keys are already loaded.

I'm not sure what it means by the keys already loaded - the way it's expressed suggests it's going to decrypt the filesystems during replication, which I don't want.

I have some encrypted filesystems and some unencrypted ones, and I just want them sent as is.  The encrypted ones should go as is (whether keys are loaded or not) and the unencrypted ones should also go unchanged.

I set `encrypted=true` to make sure the encrypted filesystems are sent encrypted as per the docs, but I'm getting an error on my unencrypted filesystems:

> PLANNING-ERROR (step 0/0, 0 B/0 B) sender filesystem is not encrypted but policy mandates encrypted send

I am not sure what to do, as the docs suggest that setting `encrypted=false` will *decrypt* my encrypted filesystems which I definitely don't want!

Is there a way to have the filesystems just replicated as is, without changing anything encryption related?

**Comment by problame** [comment, marked as answer]
The idea with `encrypted: true` is to fill in a gap in the ZFS.

Under the hood, encrypted ZFS datasets can be sent in two ways:

* on-disk encrypted blocks are put into the send stream => `zfs send -w`, works with and without key loaded
* decrypted blocks are put into the send stream => `zfs send`, fails if key is not loaded

However, the `-w` flag just means "raw" aka "send on-disk blocks as is".

So, "-w" also works against on unencrypted dataset.

That's a footgun if you don't trust the receiving side.
You want to make sure that you're really only sending encrypted blocks.

That's what zrepl's `encrypted` flag does.
It ensures that the filesystem is actually encrypted, then uses the `-w` flag.
And it refuses to do the send if the filesystem is not encrypted.

I think you're asking for mixed sends, i.e., 
- => #771 

i.e., a "don't care" mode for the `encrypted field, and probably you would use that in combination with `-w`. Follow above issue for updates if that's the case.

**Reply by Malvineous** [reply]
Thanks, yes that definitely looks like the issue.  I tried disabling the encryption option and now I get errors about "inherited key must be loaded".  I'm only replicating local fast SSDs to local slow HDDs so trust isn't an issue.

I guess as a workaround I could leave encrypted=true and just comment out the code that checks for encryption and throws the error?  Might be quicker than trying to learn Go to figure out how to add the new option properly.
